crates/namada/src/ledger/native_vp/ibc/mod.rs

```diff
@@ -209,6 +225,90 @@ where
         }
         Ok(())
     }
+
+    fn check_limits(&self, keys_changed: &BTreeSet<Key>) -> VpResult<bool> {
+        let mut tokens: Vec<&Address> = keys_changed
+            .iter()
+            .filter_map(|k| is_any_token_balance_key(k).map(|[key, _]| key))
+            .collect();
+        tokens.sort();
+        tokens.dedup();
+        for token in tokens {
+            // Limits
+            let mint_limit_key = mint_limit_key(token);
+            let mint_limit: Option<Amount> = self
+                .ctx
+                .read_pre(&mint_limit_key)
+                .map_err(Error::NativeVpError)?;
+            let throughput_limit_key = throughput_limit_key(token);
+            let throughput_limit: Option<Amount> = self
+                .ctx
+                .read_pre(&throughput_limit_key)
+                .map_err(Error::NativeVpError)?;
+            let (mint_limit, throughput_limit) =
+                match (mint_limit, throughput_limit) {
+                    (Some(ml), Some(tl)) => (ml, tl),
+                    _ => {
+                        let params: IbcParameters = self
+                            .ctx
+                            .read_pre(&params_key())
+                            .map_err(Error::NativeVpError)?
+                            .expect("Parameters should be stored");
+                        (
+                            mint_limit.unwrap_or(params.default_mint_limit),
+                            throughput_limit.unwrap_or(
+                                params.default_per_epoch_throughput_limit,
+                            ),
+                        )
+                    }
+                };
+
+            // Check the supply
+            let mint_amount_key = mint_amount_key(token);
+            let minted: Amount = self
+                .ctx
+                .read_post(&mint_amount_key)
+                .map_err(Error::NativeVpError)?
+                .unwrap_or_default();
+            if mint_limit < minted {
+                return Err(Error::RateLimit(format!(
+                    "Transfer exceeding the mint limit is not allowed: Mint \
+                     limit {mint_limit}, minted amount {minted}"
+                )));
+            }
```

Comment:
let's avoid doing storage reads directly, to facilitate isolating this kind of logic to the IBC crate. at the API level, we should only request something that implements `StorageRead` (which `ctx.pre()` does IIRC)